[veSync] New VeSync binding addition

dag81 opened the pull request:
[veSync] Initial binding contribution

This is first attempt at a pull request, to get a VeSync binding added (renamed from Levoit as multiple brands share the same protocol). This has been designed to support V2 of the protocol, as described in the readme, utilising the data from the pyvesync project. The current implementation is limited to the Levoit branded Air Purifiers and Air Humidifiers as per the readme.

Motivation and goal is that I brought a Core400S and wanted a native binding, hence a spare time project as a good starter project for learning the open-hab design many months ago now.

[Snapshot 3.3 Build of Binding is here](https://github.com/dag81/vesyncOpenhabPyPort/blob/main/manual%20builds/org.openhab.binding.vesync-3.3.0-SNAPSHOT.jar)

Within the readme are models tested and known to work, alongside the ones that should work pending user testing, should any reviewers have one to test with.

[Initial feedback and the thread of adjustments done for the devices marked as supported is here](https://community.openhab.org/t/levoit-binding/127930)

Hopefully I have not missed anything crucial 🤞 

Signed-off-by: David Goodyear <david.goodyear@gmail.com>

dag81 commented:
Sorry, I thought the changes based on what I had observed in previous PR's for new bindings would take longer, hence I closed the old one to save anyone picking it up, and starting to review while any changes were being made. As it happens it wasn't so much 🤞.

dag81 commented:
Hi all, Any chance of getting a review on this / feedback?

fwolter commented on bundles/org.openhab.binding.vesync/src/main/resources/OH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,343 @@
+
+		<config-description>
+			<parameter name="macId" type="text">
+				<context>text</context>
```
Are you sure this conext exists? It isn't documented here https://www.openhab.org/docs/developer/bindings/config-xml.html#xml-structure-for-configuration-descriptions

Please check all.

dag81 replied:
Thanks - sorry that's a good ref to bookmark :) Not sure where I pulled that from :( I've updated the bridge configuration while looking through this context request.

fwolter commented on bundles/org.openhab.binding.vesync/src/main/resources/OH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,343 @@
+
+	<channel-type id="airPurifierModeType">
+		<item-type>String</item-type>
+		<label>Operation mode</label>
```
```suggestion
		<label>Operation Mode</label>
```

dag81 replied:
Going through checking these labels now.

dag81 replied:
I've gone through and updated the case of the label's where I missed capitalisation. :)

fwolter commented on bundles/org.openhab.binding.vesync/src/main/resources/OH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,343 @@
+		<item-type>Number:Dimensionless</item-type>
+		<label>Air Quality</label>
+		<description>System representation of air quality</description>
+		<state readOnly="true" pattern="%.0f"/>
```
Display the QuantityType's unit. Please check all.
```suggestion
		<state readOnly="true" pattern="%.0f %unit%"/>
```

dag81 replied:
I've updated the pattern's to reflect the units, where I know them. Nearly all are dimensionless as I couldn't find the correct mappings, but the pattern's have been updated to show the units as reported by the external system. Hopefully this is ok.

dag81 replied:
Hi @fwolter, My apologies for the delay in jumping back on this - work and family commitments meant I had to handle those for a while. In regards to the unit's, I have just found our the air quality unit is [PM2.5 (Particulate Matter)](https://www.epa.gov/pm-pollution/particulate-matter-pm-basics) (I swear the app text must have changed to PM2.5 from PPM2.5). I have looked through the unit's and cannot see one this would map to. Is there an obvious match I'm missing otherwise it looks like this should possibly be Dimensionless with a pattern of `<state readOnly="true" pattern="%.0f PM2.5"/>`, but I just wanted to check there wasn't some match I had completely missed? Thanks again.

dag81 replied:
Changed to what I think is necessary will await feedback, if I've missed some way of mapping PM2.5 🤞 I have not.

fwolter replied:
Glad you figured out the correct units. Seems like I was wrong, too. The second P was indeed a bit misleading.

fwolter commented on bundles/org.openhab.binding.vesync/src/main/resources/OH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,343 @@
+	<channel-type id="deviceWaterLackingType">
+		<item-type>Switch</item-type>
+		<label>Water Low or Empty</label>
+		<description>Indicator if the devices water is low or empty?</description>
```
What's the purpose of the questionmark?

dag81 replied:
The questions marks are removed now - they were initially there as it was a interpreted guess as to what those variables meant until I got confirmation from users about each of the fields.

fwolter commented on bundles/org.openhab.binding.vesync/src/main/resources/OH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,343 @@
+
+	<channel-type id="deviceAutomaticStopReachTargetType">
+		<item-type>Switch</item-type>
+		<label>Stop @ Target</label>
```
```suggestion
		<label>Setpoint</label>
```

dag81 replied:
This isn't really a setpoint as I see it. A setpoint in my head is the final target temperature to be reached when say heating or cooling terminates. This is really a "Shutdown at Setpoint" as far as I understand it from the protocol. As its referred to as a "Target Humidity", I went with "Stop @ Target" to keep it aligned and the name short? The setpoint for humidity would be: Target Humidity Level. For now keeping as is, happy to change but then the other name should also be updated as well. (It could be Humidity Setpoint for Target Humidity Level, and Auto Off (maybe) instead of Stop @ Target?  The original names seem to describe the functionality however for the protocol?

fwolter replied:
OK, doesn't sound like a setpoint.

fwolter reviewed (changes requested):
Thanks for your contribution! I reviewed your code and here is my feedback.

There are some compiler warnings about null access left. You could take a look at `mvn clean install`. To fix these, you need to store the field to a local variable and do the null check on that.

dag81 commented on bundles/org.openhab.binding.vesync/README.md:
```diff
@@ -0,0 +1,308 @@
+The binding code is based on a lot of work done by other developers:
+
+- Contributors of (https://github.com/webdjoe/pyvesync) - Python interface for VeSync
+- Rene Scherer, Holger Eisold - (https://www.openhab.org/addons/bindings/surepetcare) Sure Petcare Binding for OpenHab as a reference point for the starting blocks of this code
```
I agree - I was following a structure from previous openHab module that had been through the review process. I'm working through the suggestions now - thanks :) . I'll move the credits block to the end of the read me while correcting all OpenHab to openHab)

dag81 commented on bundles/org.openhab.binding.vesync/README.md:
```diff
@@ -0,0 +1,308 @@
+
+## Thing Configuration
+
+Details for the manual configuration are below, using the OpenHab3 UI hopefully it is self-explanatory.
```
Removed :)

dag81 commented on bundles/org.openhab.binding.vesync/src/main/java/org/openhab/binding/vesync/internal/handlers/VeSyncBaseDeviceHandler.java:
```diff
@@ -0,0 +1,479 @@
+    private final Object pollConfigLock = new Object();
+
+    protected @Nullable Channel findChannelById(final String channelGroupId) {
+        // return getThing().getChannels().stream().anyMatch(x -> x.getUID().getId().equals(channelGroupId));
```
Removed - thanks for pointing out.

dag81 commented on bundles/org.openhab.binding.vesync/src/main/java/org/openhab/binding/vesync/internal/handlers/VeSyncBaseDeviceHandler.java:
```diff
@@ -0,0 +1,479 @@
+        }
+    }
+
+    @NotNull
+    protected String deviceLookupKey = MARKER_INVALID_DEVICE_KEY;
```
Thanks - I missed few others at the bottom of the handlers as well - so they have been moved along with this one to the top of the files.

dag81 commented on bundles/org.openhab.binding.vesync/src/main/java/org/openhab/binding/vesync/internal/handlers/VeSyncBaseDeviceHandler.java:
```diff
@@ -0,0 +1,479 @@
+            this.updateProperties(newProps);
+            removeChannels();
+            if (!isDeviceSupported()) {
+                updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.HANDLER_REGISTERING_ERROR,
```
Fair enough - I thought that other one was more suitable, updated as suggested to communication error.

dag81 commented on bundles/org.openhab.binding.vesync/src/main/java/org/openhab/binding/vesync/internal/handlers/VeSyncBaseDeviceHandler.java:
```diff
@@ -0,0 +1,479 @@
+    protected String deviceLookupKey = MARKER_INVALID_DEVICE_KEY;
+
+    public void configurationUpdated(Thing thing) {
+        logger.debug("DETECTED CONFIG UPDATE FOR : {}", thing);
```
Thanks good catch - that should have been removed - it was left in by mistake

dag81 commented on bundles/org.openhab.binding.vesync/src/main/java/org/openhab/binding/vesync/internal/handlers/VeSyncBaseDeviceHandler.java:
```diff
@@ -0,0 +1,479 @@
+        pollForDeviceData(lastPollResultCache);
+    }
+
+    protected void pollForDeviceData(final ExpiringCache<String> cachedResponse) {
+        // Each device should implement this to get the latest data that is not part of the meta data.
+    }
```
Agreed - I should have changed that. I may have had some debug in there previously and missed that - thank you.

dag81 commented on bundles/org.openhab.binding.vesync/src/main/java/org/openhab/binding/vesync/internal/handlers/VeSyncBaseDeviceHandler.java:
```diff
@@ -0,0 +1,479 @@
+            return EMPTY_STRING;
+        }
+
+        /*
+         * if (deviceLookupKey == null) {
+         * logger.debug("No key for addressing data");
+         * return EMPTY_STRING;
+         * }
+         */
```
Removed as suggested thanks

dag81 commented on bundles/org.openhab.binding.vesync/src/main/java/org/openhab/binding/vesync/internal/handlers/VeSyncBaseDeviceHandler.java:
```diff
@@ -0,0 +1,479 @@
+    public void updateBridgeBasedPolls(VeSyncBridgeConfiguration config) {
+    }
+
+    // Sub-classes should override this to return true, if the meta-data supports the device data read.
+    protected boolean isDeviceSupported() {
+        return false;
+    }
```
Agreed - I've updated it to be abstract.

dag81 commented on bundles/org.openhab.binding.vesync/src/main/java/org/openhab/binding/vesync/internal/handlers/VeSyncBaseDeviceHandler.java:
```diff
@@ -0,0 +1,479 @@
+        }, 1L, TimeUnit.SECONDS);
+    }
+
+    private volatile long latestReadBackMillis = 0;
```
Caught this one from your last comment on the same thing thanks.

dag81 commented on bundles/org.openhab.binding.vesync/src/main/java/org/openhab/binding/vesync/internal/handlers/VeSyncBridgeHandler.java:
```diff
@@ -0,0 +1,235 @@
+        return thing.getUID();
+    }
+
+    public VeSyncBridgeHandler(Bridge bridge, VesyncV2ApiHelper api) {
```
Scanning through now finding all of these thanks.

dag81 replied:
They should now all be adjusted, when I push the latest changes.

dag81 commented on bundles/org.openhab.binding.vesync/src/main/java/org/openhab/binding/vesync/internal/handlers/VeSyncBridgeHandler.java:
```diff
@@ -0,0 +1,235 @@
+            } catch (final AuthenticationException ae) {
+                updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR, "Check login credentials");
+                // setBackgroundScanInterval(DEFAULT_DEVICE_SCAN_DISABLED); -- Let the system keep checking in case the
+                // user updates their password externally to match openhab
```
I updated this comment to -> 
`// The background scan will keep trying to authenticate in case the users credentials are updated on the veSync servers,
 // to match the binding's configuration.`

dag81 commented on bundles/org.openhab.binding.vesync/src/main/java/org/openhab/binding/vesync/internal/handlers/VeSyncBaseDeviceHandler.java:
```diff
@@ -0,0 +1,479 @@
+        logger.debug("DETECTED CONFIG UPDATE FOR : {}", thing);
+        // Get the new addressing lookup data
+        deviceLookupKey = getValidatedIdString();
+        initialize();
```
I will cross check this and update this comment, I remember there being some behaviour it was there to work around.

dag81 replied:
This was because, when its reconfigured in the base class it doesn't call initialize. If the config has changed it effectively need's to re-pull the meta-data about the device, and start-up polling if its a known device (from the data via the protocol). Rather than call the framework method's, I've remapped the code into a new function and called that from the config change, and the initialize override. In case initialize does other bit's this should remove any un-intended behaviours of the framework, I think.

fwolter replied:
What do you mean by "reconfigured in the base class"? The framework calls `initialize()` on the Thing. If you need to do some operations in the sub classes, you can call these methods from with `initialize()`.

In your current code `configurationUpdated()` is not used.

dag81 replied:
So I had a closer look your right - it looks like dead code from testing various scenario's after removing it. I've removed this configurationUpdated() handler now.

dag81 commented on bundles/org.openhab.binding.vesync/src/main/java/org/openhab/binding/vesync/internal/handlers/VeSyncBaseDeviceHandler.java:
```diff
@@ -0,0 +1,479 @@
+        newProps.put(DEVICE_PROP_DEVICE_MAC_ID, metadata.getMacId());
+        newProps.put(DEVICE_PROP_DEVICE_NAME, metadata.getDeviceName());
+        newProps.put(DEVICE_PROP_DEVICE_TYPE, metadata.getDeviceType());
+        // newProps.put(DEVICE_PROP_DEVICE_UUID, metadata.getUuid().replace("-", ""));
```
Removed as requested.

dag81 commented on bundles/org.openhab.binding.vesync/README.md:
```diff
@@ -0,0 +1,308 @@
+
+Once the bridge is configured auto discovery will discover supported devices from the VeSync API.
+
+## Binding Configuration
+
+The binding consists of a Bridge (the API connection), and at the moment one Thing for each Air Purifier, which relates to the individual hardware devices supported. VeSync things can be configured either through the online configuration utility via discovery, or manually through a 'vesync.things' configuration file. The Bridge is not automatically discovered and must be added manually. That is b ...
+
+After adding the Bridge and setting the config, it will go ONLINE. After a short while, the discovery process for the VeSync devices will start. When supported hardware is discovered it will appear in the Inbox.
```
No problem at all I'll remove - I was going by the text I'd seen in others I was using as a reference.

dag81 commented on bundles/org.openhab.binding.vesync/.gitignore:
```diff
@@ -0,0 +1,2 @@
+target/**
+.idea/**
```
Looking through the other modules I can remove that file. It's a carry over from my local development, that shouldn't be needed in the main repo.

dag81 replied:
That files been removed

dag81 commented on bundles/org.openhab.binding.vesync/src/main/java/org/openhab/binding/vesync/internal/handlers/VeSyncBaseDeviceHandler.java:
```diff
@@ -0,0 +1,479 @@
+            return EMPTY_STRING;
+        } catch (final DeviceUnknownException e) {
+            logger.debug("Device unknown exception {}", e.getMessage());
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.HANDLER_REGISTERING_ERROR,
+                    "Check configuration details");
```
I assume you mean in the description field? I've updated the code to be: 
`"Check configuration details - "+e.getMessage()`

dag81 commented:
Hi @fwolter,

Many thanks for going through - I'm still working through each of the comments. The bit below puzzles me however:

RE: There are some compiler warnings about null access left. You could take a look at mvn clean install. To fix these, you need to store the field to a local variable and do the null check on that.

I'm missing these?  any chance you can point me to doc's about what may be causing this env not to report them? My output is below, just below my JDK details. Are you using Oracles implementation of the JDK? Are you seeing these errors in PMD processing - where mine is crashing out?

```
PS C:\GitHubRepos\dag81-openhab-addons\openhab-addons\bundles> javac -version
javac 11.0.13
PS C:\GitHubRepos\dag81-openhab-addons\openhab-addons\bundles> java --version
openjdk 11.0.13 2021-10-19 LTS
OpenJDK Runtime Environment Zulu11.52+13-CA (build 11.0.13+8-LTS)
OpenJDK 64-Bit Server VM Zulu11.52+13-CA (build 11.0.13+8-LTS, mixed mode)
```

```
PS C:\GitHubRepos\dag81-openhab-addons\openhab-addons\bundles> mvn clean spotless:apply install -pl org.openhab.binding.vesync
[INFO] Scanning for projects...
[INFO] 
[INFO] -------< org.openhab.addons.bundles:org.openhab.binding.vesync >--------
[INFO] Building openHAB Add-ons :: Bundles :: VeSync Binding 3.3.0-SNAPSHOT
[INFO] --------------------------------[ jar ]---------------------------------
[INFO] 
[INFO] --- maven-clean-plugin:3.0.0:clean (default-clean) @ org.openhab.binding.vesync ---
[INFO] Deleting C:\GitHubRepos\dag81-openhab-addons\openhab-addons\bundles\org.openhab.binding.vesync\target
[INFO] 
[INFO] --- spotless-maven-plugin:2.0.3:apply (default-cli) @ org.openhab.binding.vesync ---
[INFO] 
[INFO] --- maven-enforcer-plugin:3.0.0-M2:enforce (enforce-java) @ org.openhab.binding.vesync ---
[INFO] 
[INFO] --- directory-maven-plugin:0.3.1:directory-of (directories) @ org.openhab.binding.vesync ---
[INFO] Directory of org.openhab.addons:org.openhab.addons.reactor set to: C:\GitHubRepos\dag81-openhab-addons\openhab-addons
[INFO]
[INFO] --- spotless-maven-plugin:2.0.3:check (codestyle_check) @ org.openhab.binding.vesync ---
[INFO] 
[INFO] --- maven-dependency-plugin:3.1.1:unpack-dependencies (embed-dependencies) @ org.openhab.binding.vesync ---
[INFO] 
[INFO] --- maven-dependency-plugin:3.1.1:unpack (unpack-eea) @ org.openhab.binding.vesync ---
[INFO] Configured Artifact: org.lastnpe.eea:eea-all:2.2.1:jar
[INFO] Unpacking C:\Users\Household\.m2\repository\org\lastnpe\eea\eea-all\2.2.1\eea-all-2.2.1.jar to C:\GitHubRepos\dag81-openhab-addons\openhab-addons\bundles\org.openhab.binding.vesync\target\dependency with includes "" and excludes ""
[INFO] 
[INFO] --- karaf-maven-plugin:4.3.6:features-generate-descriptor (compile) @ org.openhab.binding.vesync ---
[INFO] Generating feature descriptor file C:\GitHubRepos\dag81-openhab-addons\openhab-addons\bundles\org.openhab.binding.vesync\target\feature\feature.xml
[INFO]  Tree listing is saved here: C:\GitHubRepos\dag81-openhab-addons\openhab-addons\bundles\org.openhab.binding.vesync\target\history\treeListing.txt

[INFO] ...done!
[INFO] Attaching features XML
[INFO]
[INFO] --- maven-resources-plugin:3.0.2:resources (default-resources) @ org.openhab.binding.vesync ---
[INFO] Using 'UTF-8' encoding to copy filtered resources.
[INFO] Copying 2 resources
[INFO] 
[INFO] --- maven-compiler-plugin:3.8.1:compile (default-compile) @ org.openhab.binding.vesync ---
[INFO] Changes detected - recompiling the module!
[INFO] 
[INFO] --- bnd-maven-plugin:6.1.0:bnd-process (default) @ org.openhab.binding.vesync ---
[INFO] 
[INFO] --- maven-resources-plugin:3.0.2:testResources (default-testResources) @ org.openhab.binding.vesync ---
[INFO] Using 'UTF-8' encoding to copy filtered resources.
[INFO] skip non existing resourceDirectory C:\GitHubRepos\dag81-openhab-addons\openhab-addons\bundles\org.openhab.binding.vesync\src\test\resources
[INFO]
[INFO] --- maven-compiler-plugin:3.8.1:testCompile (default-testCompile) @ org.openhab.binding.vesync ---
[INFO] No sources to compile
[INFO]
[INFO] --- maven-surefire-plugin:3.0.0-M5:test (default-test) @ org.openhab.binding.vesync ---
[INFO] No tests to run.
[INFO]
[INFO] --- maven-jar-plugin:3.0.2:jar (default-jar) @ org.openhab.binding.vesync ---
[INFO] Building jar: C:\GitHubRepos\dag81-openhab-addons\openhab-addons\bundles\org.openhab.binding.vesync\target\org.openhab.binding.vesync-3.3.0-SNAPSHOT.jar
[INFO] 
[INFO] --- maven-source-plugin:3.0.1:jar-no-fork (attach-sources) @ org.openhab.binding.vesync ---
[INFO] Building jar: C:\GitHubRepos\dag81-openhab-addons\openhab-addons\bundles\org.openhab.binding.vesync\target\org.openhab.binding.vesync-3.3.0-SNAPSHOT-sources.jar
[INFO] 
[INFO] --- sat-plugin:0.12.0:checkstyle (sat-all) @ org.openhab.binding.vesync ---
[INFO] 
[INFO] --- sat-plugin:0.12.0:pmd (sat-all) @ org.openhab.binding.vesync ---
[INFO] PMD version: 6.39.0
[WARNING] There are 3 PMD processing errors:
[WARNING] C:\GitHubRepos\dag81-openhab-addons\openhab-addons\bundles\org.openhab.binding.vesync\src\main\java\org\openhab\binding\vesync\internal\api\VesyncV2ApiHelper.java: PMDException: Error while parsing C:\GitHubRepos\dag81-openhab-addons\openhab-addons\bundles\org.openhab.binding.vesync\src\main\java\org\
openhab\binding\vesync\internal\api\VesyncV2ApiHelper.java
C:\GitHubRepos\dag81-openhab-addons\openhab-addons\bundles\org.openhab.binding.vesync\src\main\java\org\openhab\binding\vesync\internal\handlers\VeSyncBaseDeviceHandler.java: PMDException: Error while parsing C:\GitHubRepos\dag81-openhab-addons\openhab-addons\bundles\org.openhab.binding.vesync\src\main\java\org
\openhab\binding\vesync\internal\handlers\VeSyncBaseDeviceHandler.java
C:\GitHubRepos\dag81-openhab-addons\openhab-addons\bundles\org.openhab.binding.vesync\src\main\java\org\openhab\binding\vesync\internal\handlers\VeSyncBridgeHandler.java: PMDException: Error while parsing C:\GitHubRepos\dag81-openhab-addons\openhab-addons\bundles\org.openhab.binding.vesync\src\main\java\org\ope
nhab\binding\vesync\internal\handlers\VeSyncBridgeHandler.java
[INFO] 
[INFO] --- sat-plugin:0.12.0:spotbugs (sat-all) @ org.openhab.binding.vesync ---
[INFO] Fork Value is true
[INFO] Done SpotBugs Analysis....
[INFO] 
[INFO] --- sat-plugin:0.12.0:report (sat-all) @ org.openhab.binding.vesync ---
[INFO] Empty report will not be appended to the summary report.
[INFO] 
[INFO] --- karaf-maven-plugin:4.3.6:verify (karaf-feature-verification) @ org.openhab.binding.vesync ---
[INFO] Using repositories: https://openhab.jfrog.io/openhab/libs-release@id=openhab-release,https://openhab.jfrog.io/openhab/libs-snapshot@id=openhab-snapshot@noreleases@snapshots,https://repo1.maven.org/maven2@id=central
[INFO] Verification of feature openhab-binding-vesync/3.3.0.SNAPSHOT succeeded
[INFO] Features verified: 1, failures: 0, ignored: 0, skipped: 0
[INFO]
[INFO] --- maven-install-plugin:2.5.2:install (default-install) @ org.openhab.binding.vesync ---
[INFO] Installing C:\GitHubRepos\dag81-openhab-addons\openhab-addons\bundles\org.openhab.binding.vesync\target\org.openhab.binding.vesync-3.3.0-SNAPSHOT.jar to C:\Users\Household\.m2\repository\org\openhab\addons\bundles\org.openhab.binding.vesync\3.3.0-SNAPSHOT\org.openhab.binding.vesync-3.3.0-SNAPSHOT.jar
[INFO] Installing C:\GitHubRepos\dag81-openhab-addons\openhab-addons\bundles\org.openhab.binding.vesync\pom.xml to C:\Users\Household\.m2\repository\org\openhab\addons\bundles\org.openhab.binding.vesync\3.3.0-SNAPSHOT\org.openhab.binding.vesync-3.3.0-SNAPSHOT.pom
[INFO] Installing C:\GitHubRepos\dag81-openhab-addons\openhab-addons\bundles\org.openhab.binding.vesync\target\feature\feature.xml to C:\Users\Household\.m2\repository\org\openhab\addons\bundles\org.openhab.binding.vesync\3.3.0-SNAPSHOT\org.openhab.binding.vesync-3.3.0-SNAPSHOT-features.xml
[INFO] Installing C:\GitHubRepos\dag81-openhab-addons\openhab-addons\bundles\org.openhab.binding.vesync\target\org.openhab.binding.vesync-3.3.0-SNAPSHOT-sources.jar to C:\Users\Household\.m2\repository\org\openhab\addons\bundles\org.openhab.binding.vesync\3.3.0-SNAPSHOT\org.openhab.binding.vesync-3.3.0-SNAPSHOT
-sources.jar
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  42.836 s
[INFO] Finished at: 2022-04-02T23:51:10+01:00
[INFO] ------------------------------------------------------------------------
[INFO] Static code analysis summary report is available in:
[INFO] file:/C:/GitHubRepos/dag81-openhab-addons/openhab-addons/bundles/target/summary_report.html
`
```

dag81 commented on bundles/org.openhab.binding.vesync/src/main/java/org/openhab/binding/vesync/internal/handlers/VeSyncBaseDeviceHandler.java:
```diff
@@ -0,0 +1,479 @@
+        bridge.updateThing(this);
+
+        // Give the bridge time to build the datamaps of the devices
+        scheduler.schedule(this::pollForUpdate, 10, TimeUnit.SECONDS);
```
I didn't put in a object to store this as its a one-shot execution, not a re-occurring execution. I'll store the ScheduledFuture and cancel it out anyway should the thing be disposed, within 10 seconds.

dag81 commented on bundles/org.openhab.binding.vesync/src/main/java/org/openhab/binding/vesync/internal/handlers/VeSyncBaseDeviceHandler.java:
```diff
@@ -0,0 +1,479 @@
+    public void performReadbackPoll() {
+        final long requestSystemMillis = System.currentTimeMillis();
+        latestReadBackMillis = requestSystemMillis;
+        scheduler.schedule(() -> {
```
Same as before - I have adjusted to store and cancel if necessary in dispose. (I'm guessing if this hasn't been performed the system may have bigger issues with overloading).

dag81 commented on bundles/org.openhab.binding.vesync/src/main/resources/OH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,343 @@
+
+		<config-description>
+			<parameter name="username" type="text">
+				<context>password</context>
```
Ah someone who was testing raised this point. As it was PII data I left it as non visible, as the eye icon can expose it. However, as now two people have reacted, I'll adjust this. :)

dag81 commented on bundles/org.openhab.binding.vesync/src/main/resources/OH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,343 @@
+	xsi:schemaLocation="https://openhab.org/schemas/thing-description/v1.0.0 https://openhab.org/schemas/thing-description-1.0.0.xsd">
+
+	<bridge-type id="bridge">
+		<label>Vesync Bridge</label>
```
Thanks - All updated to VeSync - bits like this I don't know how I missed :(

dag81 commented:
Converted to draft while I continue to work through the comments.

dag81 commented on bundles/org.openhab.binding.vesync/README.md:
```diff
@@ -0,0 +1,308 @@
+Air Filtering models supported are Core200S and Core600S.
+Air Humidifier Classic 200S (Same as 300S without the nightlight from initial checks)
+
+### Restrictions / TODO
+
+1. Look at protocol mapping the V1 specification for the LV prefixed model.
+2. Look at potentially other equipment supported by the VeSync API.
+3. Add remaining write commands to complete Humidifier 600S support, if warm mode is controllable.
+4. Adjust Filtering Core200S & Core600S support based on feedback, given initial conversion.
+5. Adjust Humidifier 200S support based on feedback, given initial conversion.
```
Agreed, I'll make git hub issues in my local github against them, and I guess when its merged. I can raise issues against the binding in the main repo for anything still outstanding?

fwolter replied:
Sure!

dag81 commented on bundles/org.openhab.binding.vesync/README.md:
```diff
@@ -0,0 +1,308 @@
+
+### AirPurifier Thing
+
+| Channel                | Type                    | Description                                               | Model's Supported |  
```
Thanks its easy once you know what to search for :) All tables have been formatted - although I still need to update them, based on your other comments.

dag81 commented on bundles/org.openhab.binding.vesync/README.md:
```diff
@@ -0,0 +1,308 @@
+| **enabled**            | Switch                  | Whether the hardware device is enabled (Switched on)      | 600S, 400S, 300S  |
+| **childLock**         | Switch                  | Whether the child lock (display lock is enabled)          | 600S, 400S, 300S  |
+| **display**            | Switch                  | Whether the display is enabled (display is shown)         | 600S, 400S, 300S  |
+| **fanMode**           | String                  | The operation mode of the fan                             | 600S, 400S, 300S  |
```
Added in, with multiple rows for the delta's between the different variances between the models.

dag81 commented on bundles/org.openhab.binding.vesync/README.md:
```diff
@@ -0,0 +1,308 @@
+| **childLock**         | Switch                  | Whether the child lock (display lock is enabled)          | 600S, 400S, 300S  |
+| **display**            | Switch                  | Whether the display is enabled (display is shown)         | 600S, 400S, 300S  |
+| **fanMode**           | String                  | The operation mode of the fan                             | 600S, 400S, 300S  |
+| **manualFanSpeed**   | String                  | The speed of the fan when in manual mode                  | 600S, 400S, 300S  |
```
Thanks - good catch! (Insert a head banging against a wall image) :)

dag81 commented on bundles/org.openhab.binding.vesync/README.md:
```diff
@@ -0,0 +1,308 @@
+| **display**            | Switch                  | Whether the display is enabled (display is shown)         | 600S, 400S, 300S  |
+| **fanMode**           | String                  | The operation mode of the fan                             | 600S, 400S, 300S  |
+| **manualFanSpeed**   | String                  | The speed of the fan when in manual mode                  | 600S, 400S, 300S  |
+| **nightLightMode**   | String                  | The night lights mode                                     | 600S, 400S, 300S  |
```
This has been updated

fwolter commented:
PMD is bothering about these lines
```
VesyncManagedDevicesPage.Result.@Nullable VesyncManagedDeviceBase metadata = vesyncBridgeHandler.api
                    .getMacLookupMap().get(deviceLookupKey);
```
Propably because of the `@Nullable`. Can you rename one of these classes, so that you can add an import for it?

The null warnings I saw where eclipse complaining about redundant null checks, but these are results of `Map.get()`. You can leave it as is.

dag81 commented on bundles/org.openhab.binding.vesync/src/main/java/org/openhab/binding/vesync/internal/VeSyncHandlerFactory.java:
```diff
@@ -0,0 +1,110 @@
+    private static final Set<ThingTypeUID> SUPPORTED_THING_TYPES_UIDS = Set.of(THING_TYPE_BRIDGE,
+            THING_TYPE_AIR_PURIFIER, THING_TYPE_AIR_HUMIDIFIER);
+
+    private final VesyncV2ApiHelper api = new VesyncV2ApiHelper();
```
Hi, logically the user should only need one bridge. However this always annoyed me when I started (but was taken from existing code) - so have re-visited now  and moved the API to each bridge - now I understand the hierarchy more in openHab. The handler factory, now has a manager to notify the bridges should there be any update required of the HttpClient. (P.S I did cross check with multiple bridges after your comment, and yes it was broken should users have tried to use it in that manner). This should hopefully be resolved now.

fwolter replied:
The HttpClient is not updated during runtime. So, no need to track updates.

dag81 replied:
Thanks okay by the way I had seen it used I assumed updates did need to be tracked. (Another good to know item :) - Thank you). I've passed an Interface through so the complete factory isn't referenced so it can be used during initialisation. I didn't set it during construction as I observed the system hadn't actually got the httpClient when objects were constructed in some some tests. Hence there is a reference to where it can pull it from, and then as part of initialisation it's as the first step in the process.

dag81 commented on bundles/org.openhab.binding.vesync/README.md:
```diff
@@ -0,0 +1,308 @@
+# VeSync Binding
+
+It's current support is for the Air purifiers & Humidifer's branded as Levoit which utilise the VeSync app based on the V2 protocol.
```
Applied as suggested (I've learnt now not to use that Commit suggestion button, after it committed without the code sign-off :) - good old fun git cli required then).

dag81 commented on bundles/org.openhab.binding.vesync/README.md:
```diff
@@ -0,0 +1,308 @@
+| schedulesCount        | Number                  | The number schedules configured                           | 600S, 400S        |
+| configDisplayForever | Switch                  | Config: Whether the display will disable when not active  | 600S, 400S, 300S  |
+| configAutoMode       | String                  | Config: The mode of operation when auto is active         | 600S, 400S, 300S  |
+| configAutoRoomSize  | Number                  | Config: The room size set when auto utilises the room size| 600S, 400S, 300S  |
```
Allow `Numbers` have been updated to `Number:Dimensionless` as these were the most suitable mappings I could find when I went through. The doc's should now reflect this.

dag81 commented on bundles/org.openhab.binding.vesync/README.md:
```diff
@@ -0,0 +1,308 @@
+|------------------------|-------------------------|-------------------------------------------------------|
+| username               | String                  | The username as used in the VeSync mobile application |
+| password               | String                  | The password as used in the VeSync mobile application |
+| airPurifierPollInterval| String                  | The poll interval for air filters / humidifers        |
```
Thanks good catch - updated to number.

dag81 commented on bundles/org.openhab.binding.vesync/src/main/java/org/openhab/binding/vesync/internal/dto/VeSyncBridgeConfiguration.java:
```diff
@@ -0,0 +1,30 @@
+    public String username;
+    public String password;
+    public long airPurifierPollInterval = DEFAULT_POLL_INTERVAL_AIR_FILTERS_DEVICES;
+    public boolean backgroundDeviceDiscovery;
+    public long refreshBackgroundDeviceDiscovery = DEFAULT_REFRESH_INTERVAL_DISCOVERED_DEVICES;
```
I've updated the documentation to reflect these, along with recommended values, as due to the nature of how the binding works, these recommended values should be reasonable optimal and not overload VeSync's servers too much, while maintaining a performant system.

dag81 commented on bundles/org.openhab.binding.vesync/src/main/resources/OH-INF/binding/binding.xml:
```diff
@@ -0,0 +1,9 @@
+	xsi:schemaLocation="https://openhab.org/schemas/binding/v1.0.0 https://openhab.org/schemas/binding-1.0.0.xsd">
+
+	<name>VeSync Binding</name>
+	<description>This is the binding for VeSync.</description>
```
I've updated this description to: 
`This is the binding for the VeSync products. Currently, this supports the Levoit branded Air Purifiers and Humidifiers using the v2 protocol. ` Hopefully that's as expected.

dag81 commented on bundles/org.openhab.binding.vesync/src/main/resources/OH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,343 @@
+				<description>Password for the registered VeSync username, that allows to access the mobile application.</description>
+			</parameter>
+			<parameter name="backgroundDeviceDiscovery" type="boolean">
+				<label>Enable Background Device Scans</label>
```
I've tried to cut these down to what's required. Hopefully they are acceptable now?

dag81 commented on bundles/org.openhab.binding.vesync/src/main/java/org/openhab/binding/vesync/internal/discovery/VeSyncDiscoveryService.java:
```diff
@@ -0,0 +1,143 @@
+            properties.put(DEVICE_PROP_CONFIG_DEVICE_MAC, apMeta.getMacId());
+            properties.put(DEVICE_PROP_CONFIG_DEVICE_NAME, apMeta.getDeviceName());
+            return DiscoveryResultBuilder.create(new ThingUID(THING_TYPE_AIR_PURIFIER, bridgeUID, deviceUUID))
+                    .withLabel(apMeta.getDeviceName()).withBridge(bridgeUID).withProperties(properties).build();
```
I spent ages trying to get that working correctly prior to seeing your comment - thank you :) Adjusted now the representation prop's to filter out as expected :) Both updated and checked.

dag81 commented on bundles/org.openhab.binding.vesync/src/main/java/org/openhab/binding/vesync/internal/discovery/VeSyncDiscoveryService.java:
```diff
@@ -0,0 +1,143 @@
+        // Temporary until refactoring
+        bridgeHandler.getAirHumidifiersMetadata().map(apMeta -> {
+            final Map<String, Object> properties = new HashMap<>(6);
+            // final String deviceUUID = apMeta.getUuid().replace("-", "");
```
Removed during addressing of last comment :)

dag81 commented on bundles/org.openhab.binding.vesync/README.md:
```diff
@@ -0,0 +1,308 @@
+| Name                   | Type                    | Description                                                         |
+|------------------------|-------------------------|---------------------------------------------------------------------|
+| deviceName             | String                  | The name given to the device under Settings -> Device Name          |
+| macId                  | String                  | The mac for the device under Settings -> Device Info -> MAC Address |
```
Adjusted to hopefully the correct location.

dag81 commented on bundles/org.openhab.binding.vesync/README.md:
```diff
@@ -0,0 +1,308 @@
+| **mistLevel**            | Number:Dimensionless    | The current mist level set (1-2) or (1-3)                  | 200S, Dual200S, 300S, 600S |
+| **humidifierMode**       | String                  | The current mode of operation [auto,sleep]                 | 200S, Dual200S, 300S, 600S |
+| **nightLightMode**       | String                  | The night light mode [on,dim,off]                          | 300S                       |
+| **configTargetHumidity** | Number:Dimensionless    | Config: What the target humidity is set to reach           | 200S, Dual200S, 300S, 600S |
```
This pairs up with the other comment about the stop at target. I've renamed the channels to humiditySetpoint, and stopAtHumiditySetpoint. The config bit isn't really suitable, but a hangover of not having the units directly to work with, so it was a raw protocol mapping / interpretation. So adjusted and hopefully good now :)

dag81 commented on bundles/org.openhab.binding.vesync/src/main/java/org/openhab/binding/vesync/internal/VeSyncHandlerFactory.java:
```diff
@@ -0,0 +1,110 @@
+        return null;
+    }
+
+    private ThingUID getDeviceUID(ThingTypeUID thingTypeUID, @Nullable ThingUID thingUID, Configuration configuration,
+            @Nullable ThingUID bridgeUID) {
+        if (thingUID != null) {
+            return thingUID;
+        } else {
+            if (bridgeUID != null) {
+                return new ThingUID(thingTypeUID, configuration.get(DEVICE_MAC_ID).toString(), bridgeUID.getId());
+            } else {
+                return new ThingUID(thingTypeUID, configuration.get(DEVICE_MAC_ID).toString());
+            }
+        }
+    }
```
I just updated this. I think it was for the trick I missed withRepresentation before, that you pointed out. The software run's fine without it, so I've removed this getDeviceUID override. Thanks for the shout out on withRepresentation before :)

dag81 commented:
Hi @fwolter, 

Many thanks for going through all of the code, some bits especially I wondered why I missed it. It is however a side project, so ... I've adjusted for your comments that you raised, I assume with the openHab PR's you resolve the issues once happy (if its standard work place procedures :) )? or should I go through and mark as resolved as I think they are handled?

(Apologies for all the 👍 I needed a way to visually navigate all the comment's I had addressed - or thought I had).

Finally sorry, I don't understand what you mean to resolve the PMD issues. I hate any errors, so would like to get rid of them. I just cant quite see what you mean at the moment in the last comment? Could you re-phrase for me please - sorry.

Many thanks again

fwolter commented:
You can resolve the issues when you finsihed working on it. I will re-open any, if necessary.

I meant the errors you get during build:
```
[INFO] --- sat-plugin:0.12.0:pmd (sat-all) @ org.openhab.binding.vesync ---
[INFO] PMD version: 6.39.0
[WARNING] There are 3 PMD processing errors:
[WARNING] C:\GitHubRepos\dag81-openhab-addons\openhab-addons\bundles\org.openhab.binding.vesync\src\main\java\org\openhab\binding\vesync\internal\api\VesyncV2ApiHelper.java: PMDException: Error while parsing C:\GitHubRepos\dag81-openhab-addons\openhab-addons\bundles\org.openhab.binding.vesync\src\main\java\org\
openhab\binding\vesync\internal\api\VesyncV2ApiHelper.java
C:\GitHubRepos\dag81-openhab-addons\openhab-addons\bundles\org.openhab.binding.vesync\src\main\java\org\openhab\binding\vesync\internal\handlers\VeSyncBaseDeviceHandler.java: PMDException: Error while parsing C:\GitHubRepos\dag81-openhab-addons\openhab-addons\bundles\org.openhab.binding.vesync\src\main\java\org
\openhab\binding\vesync\internal\handlers\VeSyncBaseDeviceHandler.java
C:\GitHubRepos\dag81-openhab-addons\openhab-addons\bundles\org.openhab.binding.vesync\src\main\java\org\openhab\binding\vesync\internal\handlers\VeSyncBridgeHandler.java: PMDException: Error while parsing C:\GitHubRepos\dag81-openhab-addons\openhab-addons\bundles\org.openhab.binding.vesync\src\main\java\org\ope
nhab\binding\vesync\internal\handlers\VeSyncBridgeHandler.java
```
They prevent running checkstyle and so on. So, there might be other hidden warnings.

dag81 commented:
Hi @fwolter,

I got it now - PMD didn't like the nested inner classes :) Thanks again - I've removed the draft status now.

dag81 commented:
@fwolter Thanks again for the time looking at this! :) I've addressed your latest two comments I believe. I couldn't see the button to re-request a review so I assume its still open with yourself? Hopefully the changes make sense 🤞

dag81 commented on bundles/org.openhab.binding.vesync/README.md:
```diff
@@ -0,0 +1,303 @@
+| backgroundDeviceDiscovery        | Switch | Should the system scan periodically for new devices       | ON                 |
+| refreshBackgroundDeviceDiscovery | Number | Frequency (seconds) of scans for new new devices          | 120                |
+
+* Note Air PPM Levels don't usually change quickly - 60s seems reasonable if openHab is controlling it and your don't want near instant feedback of physical interactions with the devices.
```
Fix two instances I found in the README.md file from openHab -> openHAB :)

dag81 commented on bundles/org.openhab.binding.vesync/src/main/java/org/openhab/binding/vesync/internal/handlers/VeSyncBaseDeviceHandler.java:
```diff
@@ -0,0 +1,510 @@
+            // Check if the device name can be matched to a single device
+            if (deviceName != null) {
+                final String[] matchedMacIds = vesyncBridgeHandler.api.getMacLookupMap().values().stream()
+                        .filter(x -> deviceName.equals(x.deviceName)).map(x -> x.macId).toArray(String[]::new); // .collect();//.toArray(String[]::new)
```
Absolutely - done

dag81 commented on bundles/org.openhab.binding.vesync/src/main/java/org/openhab/binding/vesync/internal/handlers/VeSyncDeviceAirHumidifierHandler.java:
```diff
@@ -0,0 +1,348 @@
+        updateState(DEVICE_CHANNEL_WATER_TANK_LIFTED, OnOffType.from(humidifierStatus.result.result.waterTankLifted));
+        updateState(DEVICE_CHANNEL_STOP_AT_TARGET,
+                OnOffType.from(humidifierStatus.result.result.automaticStopReachTarget));
+        updateState(DEVICE_CHANNEL_HUMIDITY, new DecimalType(humidifierStatus.result.result.humidity));
```
Ah, thank you. Let me go through and check them all.

dag81 commented on bundles/org.openhab.binding.vesync/src/main/resources/OH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,341 @@
+		<item-type>Number:Dimensionless</item-type>
+		<label>Air Quality PPM2.5</label>
+		<description>Indicator of current air quality</description>
+		<state readOnly="true" pattern="%.0f PPM 2.5"/>
```
@fwolter Could do with further input here - sorry - and thanks again for all the time looking through this!!! I was just thinking about this one, as it jogged memories from when this was done last year. So the VeSync app and physical devices report the air quality as a unit from 0 - 999 in PM 2.5 which I assume is particles per 2.5 million. Really for the units to be PPM, the reported value would have to be manipulated to the corresponding scale. With my user hat on, its much simpler to write rules based on the units / app's display, also as a user I would think the openHAB interface is incorrect if it was giving a different displayed number to what the app and unit reports as coming from the device? I'm not sure whether it should be adjusted or not - but I know from using these unit's myself it would be much harder to figure out the rule's /  conditions if the numbers were scaled to ppm? I'd personally probably opt for not manipulating the reported numbers to a scaled PPM, just to save people thinking the openHAB interface is wrong, and by using the devices unit's it simplifies the creation of rules, as I know what a PM2.5 reading of 15 is but wouldn't know the corresponding PPM value easily. If the numbers are not scaled then a ppm unit isn't correct as its got the 2.5 scaler attribute as part of it?

fwolter replied:
PPM2.5 is a common value to measure fine particles. It means 2.5µm-sized particles per million. See https://www.health.ny.gov/environmental/indoors/air/pmq_a.htm

So, you can take the VeSync value as is and simply use ppm as the unit.

dag81 replied:
Units corrected now :)

dag81 commented on bundles/org.openhab.binding.vesync/src/main/java/org/openhab/binding/vesync/internal/handlers/VeSyncBaseDeviceHandler.java:
```diff
@@ -0,0 +1,510 @@
+        } catch (final DeviceUnknownException e) {
+            logger.debug("Device unknown exception {}", e.getMessage());
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.HANDLER_REGISTERING_ERROR,
+                    "Check configuration details - " + e.getMessage());
```
I have searched the code for all` HANDLER_REGISTERING_ERROR` and updated to `COMMUNICATION_ERROR`, apologies I know you pointed this one out before. The log's have been removed as now included as part of the status message, like you suggested.

dag81 commented on bundles/org.openhab.binding.vesync/src/main/java/org/openhab/binding/vesync/internal/handlers/VeSyncBaseDeviceHandler.java:
```diff
@@ -0,0 +1,510 @@
+                throw new DeviceUnknownException("Missing client");
+            }
+        } catch (AuthenticationException e) {
+            logger.debug("Auth exception {}", e.getMessage());
```
So in regard's to the Authentication Exception. Generally they should never actually hit this, as the bridge should auth fail out on its check, or its heart beats to read the meta-data. If they auth fail the bridge will go offline, which in turn as I understand it from openHAB's framework will ripple to the sub-ordinate devices under that bridge. The only point they could hit this temporarily is if they changed the password for the VeSync servers, then it the bridge would pick it up at it's next meta-data scan - which would force everything off-line as the authentication is failing. The offline status is driven primarily out-side of this use case based on what VeSync server's detects. So if the meta-data say's a device is off-line then that explicit device is marked as off-line until the next meta-data read shows it as back online. (If any device is off-line the meta-data reading increases in frequency, to minimise the lag in detecting a device is back online). So I guess the short answer is the bridge handles auth issues, and the device's generally handle their online status as reported by the state reported by VeSync's cloud API.

fwolter replied:
OK, then it's good to keep it as is.

dag81 commented on bundles/org.openhab.binding.vesync/src/main/java/org/openhab/binding/vesync/internal/VeSyncHandlerFactory.java:
```diff
@@ -0,0 +1,97 @@
+        return null;
+    }
+
+    @Override
+    public @Nullable Thing createThing(final ThingTypeUID thingTypeUID, final Configuration configuration,
+            final @Nullable ThingUID thingUID, final @Nullable ThingUID bridgeUID) {
+        // Use the specific Handler Factory if required
+        // otherwise fallback to the default
+        if (VeSyncDeviceAirPurifierHandler.SUPPORTED_THING_TYPES_UIDS.contains(thingTypeUID)
+                || VeSyncDeviceAirHumidifierHandler.SUPPORTED_THING_TYPES_UIDS.contains(thingTypeUID)
+                || VeSyncHandlerFactory.SUPPORTED_THING_TYPES_UIDS.contains(thingTypeUID)) {
+            return super.createThing(thingTypeUID, configuration, thingUID, bridgeUID);
+        } else {
+            throw new IllegalArgumentException(
+                    "The thing type " + thingTypeUID + " is not supported by the VeSync binding.");
+        }
+    }
```
Sorry @fwolter, I think I'm missing something here. So the purpose of this block is to cross check the ID's reported by the devices and confirm that one of the specialised classes in the binding can provide the functionality for interacting with that device. It's checking list's in each class based on the devices reported id code whether a instance will be able to interact with it. If it isn't an exception is thrown so a ticket / issue can be raised to add support if the relevant reverse engineering is available in pyvesync. How would the framework know to check those list's and block creating a thing, if it isn't supported, should the super class instance run?

fwolter replied:
The framework queries your factory code before creating a new Thing:
```
    @Override
    public boolean supportsThingType(ThingTypeUID thingTypeUID) {
        return SUPPORTED_THING_TYPES_UIDS.contains(thingTypeUID);
    }
```
If this returns false, the framework rejects the creation. So, no need to check that yourself.

dag81 replied:
Thank you - i'll adjust

fwolter commented on bundles/org.openhab.binding.vesync/src/main/resources/OH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,341 @@
+
+	</thing-type>
+
+	<thing-type id="AirHumidifier">
```
Thing type IDs need to start with a lower case. Same for the purifier.
```suggestion
	<thing-type id="airHumidifier">
```

dag81 replied:
Thank you 👍 The thing-type id's have been converted to lower case, code references and documentation updated.

fwolter reviewed (approved):
LGTM
Now, you could add your binding's logo to the openHAB website. See https://www.openhab.org/docs/developer/bindings/#add-your-binding-s-logo-to-the-openhab-website
